### PR TITLE
Deprecate and remove unnecessary DefaultConfiguration methods

### PR DESCRIPTION
**Pull request by tresat**
- Deprecate the configuration.getAll method
- Remove internal isIncubating and isUsageMutable methods

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 1]
```diff
@@ -138,6 +138,15 @@ Some of them still registers conventions for backwards compatibility.
 Registering conventions does not emit a deprecation warning yet to provide a migration window.
 Future Gradle versions will do.
 
+[[deprecated_configuration_get_all]]
+==== `link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[CompileOptions]` method deprecations
```
Comment (jvandort) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:142]:
```suggestion
==== `link:{javadocPath}/org/gradle/api/artifacts/Configuration.html[Configuration]` method deprecations
```

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy [hunk 2]
```diff
@@ -180,7 +180,6 @@ class DefaultConfigurationContainerSpec extends Specification {
         then:
         detached.name == "detachedConfiguration1"
         detached.getAll() == [detached] as Set
-        detached.getHierarchy() == [detached] as Set
```
Comment (jvandort) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy:183]: Why remove this line?

**Review by jvandort — APPROVED** [review]

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy (hunk 2 above)
Comment (tresat) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainerSpec.groovy:183]: Must have been accidental.  I'll restore it.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**